staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go

```diff
@@ -44,3 +46,23 @@ func ValidateListOptions(options *internalversion.ListOptions) field.ErrorList {
 	}
 	return allErrs
 }
+
+func validateWatchOptions(options *internalversion.ListOptions, isWatchListFeatureEnabled bool) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if !options.Watch {
+		return allErrs
+	}
+	if match := options.ResourceVersionMatch; len(match) > 0 {
+		if !isWatchListFeatureEnabled {
+			allErrs = append(allErrs, field.Forbidden(field.NewPath("resourceVersionMatch"), "resourceVersionMatch is forbidden for watch"))
+			return allErrs
+		}
+		if match != metav1.ResourceVersionMatchMostRecent {
+			allErrs = append(allErrs, field.NotSupported(field.NewPath("resourceVersionMatch"), match, []string{string(metav1.ResourceVersionMatchMostRecent)}))
+		}
+		if match == metav1.ResourceVersionMatchMostRecent && !options.AllowWatchBookmarks {
```

Comment:
Shouldn't we force watch bookmarks even if the "MostRecent" param is not set?

Reply:
ping

Reply:
I don't think so. Why we should force clients to register for progress notification? 

Here we say that if you request `MostRecent` then you also need to understand bookmarks events.
I think we should also check if the feature flag for bookmarks was enabled.

Reply:
oh, `WatchBookmark` is `LockToDefault` so we are good.

Reply:
OK - I initially thought that we may want to support "list via watch" also for "NotOlderThan".
If we limit ourselves to MostRecent - then it's fine.

But I'm still chewing if we don't want to support the "NotOlderThan" for  "list via watch".


In other words - I'm thinking (I'm far from being convinced at this point, but I'm thinking) about extending ListOptions with the new field:
  `SendInitialEvents bool`

Default to false would give us backward compatibility.
If explicitly set to true (for watch requests) it would basically be sending the initial watch events according to the existing semantic of ResourceVersionMatch:
- NotOlderThan (you get bookmark when we pass the given RV)
- Exact (you get bookmark when we pass then "current etcd RV")

Reply:
You may want to ask @deads2k for his opinion which API he would prefer, I'm not an API expert.

Reply:
for the LIST we could also use `MostRecent` in that case we would wait until the cache has caught up and only then send initial data (without the bookmark event). I think that would fix the long-standing issue with consistency

Reply:
i think we're mixing two things here with this API.

I would like to allow passing the RV myself too. In other words, I would like to be able to say that instead of using MostRecent, I would like to use watch-based list but pass that RV myself (e.g. explicitly pass 100).

So I still think that an explicitly `SendInitialEvents bool` field might be a better option.

Reply:
To summarise our discussion:

We have to use `resourceVersionMatch`. 

Passing a new param (i.e. '/api/v1/namespaces/kube-system/pods?abc=xyz&watch=true') doesn't ensure backward compatibility. We won't get a 422 response. While we could patch prior versions of k8s doing so for aggregated APIs is impossible.

Given the above we have at least two options:

Option 1: Define a new resourceVersionMatch option (`MostRecent`) for WATCH requests. Allow to use it with **any** RV value.

Option 2: Define a new resourceVersionMatch option (`MostRecent`) for WATCH requests. Allow to use it **only** with RV=0 | RV="". This would allow to use it only for initial requests.


My preference is towards the `Option 1`. However, before we merge this PR I'd like to gather some feedback.

Reply:
@deads2k what is your opinion/preference?

Reply:
> OK - I initially thought that we may want to support "list via watch" also for "NotOlderThan".

I agree with this sentiment: I expected this to work as well.  `SendInitialEvents` seems to more cleanly describe what we want to happen.  That would allow the following combinations (correct me if I'm mistaken)

1. sendInitialEvent and rv>0 - this would provide the initial state and when we reach the desired RV, the synthetic bookmark (this must be recognizably different) will let you know when you are done with the list.
2. sendInitialEvent and rv=0 - when the initial state has been replayed as far as we have available, the synthetic bookmark will let you know when you are done with the list.
3. sendInitialEvent and rv="" - when the initial state has been replayed as far as etcd has progressed (MostRecent), the synthetic bookmark will let you know when you are done with the list.

It appears to me that the new ResourceVersionMatch value is being used here to allow old servers to indicate to clients that they do not support the option the client is requesting.  We have historically managed this by explicitly rejecting new options in older version.  I remember a PR doing that same thing for `DryRun`.  The choice of adding a ResourceVersionMatch is more immediately convenient, but less practical for numerous features over time as the temptation to continue overloading enums or existing options grows.  See https://github.com/kubernetes/kubernetes/pull/63557 for an example.

Reply:
> 1. sendInitialEvent and rv>0 - this would provide the initial state and when we reach the desired RV, the synthetic bookmark (this must be recognizably different) will let you know when you are done with the list.

why we would like to provide the initial state in this case? I think in that case we should only give you events that happened after the provided RV. I think this would scale better than replying the initial events the client has already seen.

> It appears to me that the new ResourceVersionMatch value is being used here to allow old servers to indicate to clients that they do not support the option the client is requesting

yes, we don't want to provide any compatibilities for the aggregated APIs?

Reply:
Thanks @deads2k for a nice summary - it exactly reflects my thinking (although I wasn't sure what's the recommended path for doing that from the API perspective).

> why we would like to provide the initial state in this case? I think in that case we should only give you events that happened after the provided RV. I think this would scale better than replying the initial events the client has already seen.

There are two cases:
- you specify the new `SendInitialEvents` parameter - then I want to see initial events. The usecase for that is that e.g. my reflector went out of history window so I can't resume my watch and need to relist. But the only think I need to ensure is to not go back in time, so I'm fine with slightly delayed data - i.e. NotOlderThan with rv=X
- you don't specify the new `SendInitialEvents` parameter - then I just don't want to see the initial events.

So as I originally pointed out - I'm actually with David wrt it.

> yes, we don't want to provide any compatibilities for the aggregated APIs?

Thinking about this more - maybe it's not strictly needed immediately? We're generally saying that you should first upgrade your apiservers and only then upgrade clients.
What if we would just:
- introduce the new option now (and reject the request if the field is not supported (based on feature gate)
- not turn on this feature in any of our libraries by default for the next N releases (you can only enable it explicitly if you know what you're doing, e.g. if you're using built-in APIs).
Would that be an acceptable guard?

Reply:
okay, I've updated the PR to reflect the above, please check if this is what we want.